File: src/ir.rs
// Computation graph IR — arena-based DAG.
// Each NodeId is an index into Graph::nodes; the arena is implicitly topologically
// ordered because nodes can only reference earlier indices.

pub type NodeId = usize;

#[derive(Clone, Debug, PartialEq)]
pub enum Op {
    // Leaves
    Input(usize),       // index into the runtime input slice
    Const(f64),         // scalar constant (broadcast to shape on eval)

    // Elementwise binary
    Add,
    Sub,
    Mul,
    Div,

    // Elementwise unary
    Neg,
    Relu,
    Step,               // Heaviside: 1 if x>0 else 0  (relu backward)
    Exp,
    Log,
    Tanh,
    Sigmoid,

    // Linear algebra
    MatMul,
    Transpose,          // 2-D transpose

    // Shape / broadcast
    Broadcast(Vec<usize>),   // broadcast input to given shape
    Reshape(Vec<usize>),
    Sum,                // reduce all elements to scalar
    SumAxis(usize),     // reduce along one axis

    // Placeholders — filled in later phases
    Softmax,
    CrossEntropy,       // inputs: logits, labels
    Conv2d { stride: usize, pad: usize },
    MaxPool { kernel: usize, stride: usize },
    BatchNorm,
}

#[derive(Clone, Debug)]
pub struct Node {
    pub op: Op,
    pub inputs: Vec<NodeId>,
    pub shape: Vec<usize>,    // output shape; [] = scalar
}

#[derive(Default, Debug)]
pub struct Graph {
    pub nodes: Vec<Node>,
    /// NodeIds of the Input leaves, in declaration order. inputs[i] is the node
    /// for runtime input slot i. Avoids scanning the arena to count inputs.
    pub inputs: Vec<NodeId>,
}

impl Graph {
    // ── primitive push ────────────────────────────────────────────────────
    pub fn push(&mut self, op: Op, inputs: Vec<NodeId>, shape: Vec<usize>) -> NodeId {
        self.nodes.push(Node { op, inputs, shape });
        self.nodes.len() - 1
    }

    // ── leaves ────────────────────────────────────────────────────────────
    pub fn input(&mut self, shape: Vec<usize>) -> NodeId {
        let idx = self.inputs.len();
        let id = self.push(Op::Input(idx), vec![], shape);
        self.inputs.push(id);
        id
    }

    pub fn constant(&mut self, val: f64, shape: Vec<usize>) -> NodeId {
        self.push(Op::Const(val), vec![], shape)
    }

    // ── elementwise binary ────────────────────────────────────────────────
    // All binary ops follow numpy broadcasting: operands are broadcast to a
    // common shape via explicit Broadcast nodes (XLA/HLO style — the IR never
    // broadcasts implicitly), then the op runs on equal shapes.
    pub fn add(&mut self, a: NodeId, b: NodeId) -> NodeId { self.binop(Op::Add, a, b) }
    pub fn sub(&mut self, a: NodeId, b: NodeId) -> NodeId { self.binop(Op::Sub, a, b) }
    pub fn mul(&mut self, a: NodeId, b: NodeId) -> NodeId { self.binop(Op::Mul, a, b) }
    pub fn div(&mut self, a: NodeId, b: NodeId) -> NodeId { self.binop(Op::Div, a, b) }

    fn binop(&mut self, op: Op, a: NodeId, b: NodeId) -> NodeId {
        let out = broadcast_shapes(&self.nodes[a].shape, &self.nodes[b].shape);
        let a = self.broadcast_to(a, &out);
        let b = self.broadcast_to(b, &out);
        self.push(op, vec![a, b], out)
    }

    /// Insert a Broadcast node iff `a`'s shape differs from `target`.
    pub fn broadcast_to(&mut self, a: NodeId, target: &[usize]) -> NodeId {
        if self.nodes[a].shape == target {
            a
        } else {
            self.broadcast(a, target.to_vec())
        }
    }

    pub fn neg(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Neg, vec![a], shape)
    }

    // ── elementwise unary ─────────────────────────────────────────────────
    pub fn relu(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Relu, vec![a], shape)
    }

    pub fn step(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Step, vec![a], shape)
    }

    pub fn exp(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Exp, vec![a], shape)
    }

    pub fn log(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Log, vec![a], shape)
    }

    pub fn tanh(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Tanh, vec![a], shape)
    }

    pub fn sigmoid(&mut self, a: NodeId) -> NodeId {
        let shape = self.nodes[a].shape.clone();
        self.push(Op::Sigmoid, vec![a], shape)
    }

    // ── linear algebra ────────────────────────────────────────────────────
    /// matmul: a is [M,K], b is [K,N] → [M,N]
    pub fn matmul(&mut self, a: NodeId, b: NodeId) -> NodeId {
        let sa = &self.nodes[a].shape;
        let sb = &self.nodes[b].shape;
        assert_eq!(sa.len(), 2, "matmul expects 2-D inputs");
        assert_eq!(sb.len(), 2, "matmul expects 2-D inputs");
        assert_eq!(sa[1], sb[0], "matmul inner-dim mismatch: {}x{} vs {}x{}", sa[0], sa[1], sb[0], sb[1]);
        let shape = vec![sa[0], sb[1]];
        self.push(Op::MatMul, vec![a, b], shape)
    }

    pub fn transpose(&mut self, a: NodeId) -> NodeId {
        let sa = &self.nodes[a].shape;
        assert_eq!(sa.len(), 2, "transpose expects 2-D input");
        let shape = vec![sa[1], sa[0]];
        self.push(Op::Transpose, vec![a], shape)
    }

    // ── reductions ────────────────────────────────────────────────────────
    pub fn sum(&mut self, a: NodeId) -> NodeId {
        self.push(Op::Sum, vec![a], vec![]) // scalar
    }

    pub fn sum_axis(&mut self, a: NodeId, axis: usize) -> NodeId {
        let mut shape = self.nodes[a].shape.clone();
        shape.remove(axis);
        self.push(Op::SumAxis(axis), vec![a], shape)
    }

    // ── shape ─────────────────────────────────────────────────────────────
    pub fn broadcast(&mut self, a: NodeId, shape: Vec<usize>) -> NodeId {
        self.push(Op::Broadcast(shape.clone()), vec![a], shape)
    }

    pub fn reshape(&mut self, a: NodeId, shape: Vec<usize>) -> NodeId {
        self.push(Op::Reshape(shape.clone()), vec![a], shape)
    }

    // ── shape helpers ─────────────────────────────────────────────────────
    pub fn shape_of(&self, id: NodeId) -> &[usize] {
        &self.nodes[id].shape
    }

    pub fn num_inputs(&self) -> usize {
        self.inputs.len()
    }
}

/// numpy broadcasting: align shapes from the right; each dim must be equal, or
/// one of them 1 (or missing). Result dim is the max. Panics on incompatible.
pub fn broadcast_shapes(a: &[usize], b: &[usize]) -> Vec<usize> {
    let n = a.len().max(b.len());
    let mut out = vec![0usize; n];
    for i in 0..n {
        // right-aligned index into each operand (1 if the dim is absent)
        let da = if i + a.len() < n { 1 } else { a[i + a.len() - n] };
        let db = if i + b.len() < n { 1 } else { b[i + b.len() - n] };
        out[i] = if da == db {
            da
        } else if da == 1 {
            db
        } else if db == 1 {
            da
        } else {
            panic!("incompatible broadcast shapes {:?} vs {:?}", a, b);
        };
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_simple_graph() {
        let mut g = Graph::default();
        let w = g.input(vec![2, 3]);
        let x = g.input(vec![3, 1]);
        let wx = g.matmul(w, x);
        assert_eq!(g.nodes[wx].shape, vec![2, 1]);
    }

    #[test]
    fn broadcast_shape_rules() {
        assert_eq!(broadcast_shapes(&[2, 3], &[3]), vec![2, 3]); // prepend
        assert_eq!(broadcast_shapes(&[2, 1], &[2, 3]), vec![2, 3]); // stretch size-1
        assert_eq!(broadcast_shapes(&[], &[4]), vec![4]); // scalar
        assert_eq!(broadcast_shapes(&[5, 1, 4], &[3, 1]), vec![5, 3, 4]);
    }

    #[test]
    #[should_panic]
    fn broadcast_incompatible_panics() {
        broadcast_shapes(&[3], &[4]); // 3 vs 4, neither is 1
    }

    #[test]
    #[should_panic]
    fn matmul_shape_mismatch_panics() {
        let mut g = Graph::default();
        let a = g.input(vec![2, 3]);
        let b = g.input(vec![4, 1]); // inner dim mismatch
        g.matmul(a, b);
    }
}

File: src/ir.cpp
#include "minijax/ir.hpp"
#include <algorithm>
#include <sstream>

namespace minijax {

const char* op_kind_name(OpKind k) {
    switch (k) {
        case OpKind::Input: return "Input";
        case OpKind::Const: return "Const";
        case OpKind::Add: return "Add";
        case OpKind::Sub: return "Sub";
        case OpKind::Mul: return "Mul";
        case OpKind::Div: return "Div";
        case OpKind::Neg: return "Neg";
        case OpKind::MatMul: return "MatMul";
        case OpKind::Relu: return "Relu";
        case OpKind::Step: return "Step";
        case OpKind::Tanh: return "Tanh";
        case OpKind::Sigmoid: return "Sigmoid";
        case OpKind::Exp: return "Exp";
        case OpKind::Log: return "Log";
        case OpKind::Sqrt: return "Sqrt";
        case OpKind::Abs: return "Abs";
        case OpKind::Sum: return "Sum";
        case OpKind::SumAxis: return "SumAxis";
        case OpKind::Transpose: return "Transpose";
        case OpKind::Broadcast: return "Broadcast";
        case OpKind::Reshape: return "Reshape";
        case OpKind::Softmax: return "Softmax";
        case OpKind::CrossEntropy: return "CrossEntropy";
        case OpKind::Conv2d: return "Conv2d";
        case OpKind::Im2Col: return "Im2Col";
        case OpKind::Col2Im: return "Col2Im";
        case OpKind::MaxPool: return "MaxPool";
        case OpKind::MaxPoolBackward: return "MaxPoolBackward";
        case OpKind::BatchNorm: return "BatchNorm";
    }
    return "?";
}

NodeId Graph::push_raw(Node n) {

    NodeId new_id = nodes_.size();
    for (NodeId in : n.inputs) {
        if (in >= new_id) {
            throw std::invalid_argument("Graph::push_raw: node input references a non-earlier index "
                                         "(violates topological-arena invariant)");
        }
    }
    nodes_.push_back(std::move(n));
    return new_id;
}

NodeId Graph::input(std::vector<size_t> shape) {
    Node n;
    n.op = OpKind::Input;
    n.shape = std::move(shape);
    n.input_slot = inputs_.size();
    NodeId id = push_raw(std::move(n));
    inputs_.push_back(id);
    return id;
}

NodeId Graph::constant(double value, std::vector<size_t> shape) {
    Node n;
    n.op = OpKind::Const;
    n.const_value = value;
    n.shape = std::move(shape);
    return push_raw(std::move(n));
}


static std::optional<std::vector<size_t>> broadcast_shape(const std::vector<size_t>& a,
                                                            const std::vector<size_t>& b) {
    size_t rank = std::max(a.size(), b.size());
    std::vector<size_t> out(rank);
    for (size_t i = 0; i < rank; ++i) {
        size_t ad = (i < rank - a.size()) ? 1 : a[i - (rank - a.size())];
        size_t bd = (i < rank - b.size()) ? 1 : b[i - (rank - b.size())];
        if (ad == bd) out[i] = ad;
        else if (ad == 1) out[i] = bd;
        else if (bd == 1) out[i] = ad;
        else return std::nullopt;
    }
    return out;
}

NodeId Graph::broadcast_to(NodeId a, std::vector<size_t> shape) {
    Node n;
    n.op = OpKind::Broadcast;
    n.inputs = {a};
    n.target_shape = shape;
    n.shape = shape;
    return push_raw(std::move(n));
}

NodeId Graph::binop(OpKind kind, NodeId a, NodeId b) {


    std::vector<size_t> sa = nodes_[a].shape;
    std::vector<size_t> sb = nodes_[b].shape;
    NodeId lhs = a, rhs = b;
    if (sa != sb) {


        auto target = broadcast_shape(sa, sb);
        if (!target) {
            std::ostringstream os;
            os << "Graph::binop(" << op_kind_name(kind) << "): incompatible shapes for broadcast";
            throw std::invalid_argument(os.str());
        }
        if (sa != *target) lhs = broadcast_to(a, *target);
        if (sb != *target) rhs = broadcast_to(b, *target);
    }
    Node n;
    n.op = kind;
    n.inputs = {lhs, rhs};
    n.shape = nodes_[lhs].shape;
    return push_raw(std::move(n));
}

NodeId Graph::add(NodeId a, NodeId b) { return binop(OpKind::Add, a, b); }
NodeId Graph::sub(NodeId a, NodeId b) { return binop(OpKind::Sub, a, b); }
NodeId Graph::mul(NodeId a, NodeId b) { return binop(OpKind::Mul, a, b); }
NodeId Graph::div(NodeId a, NodeId b) { return binop(OpKind::Div, a, b); }

NodeId Graph::neg(NodeId a) {
    Node n; n.op = OpKind::Neg; n.inputs = {a}; n.shape = nodes_[a].shape;
    return push_raw(std::move(n));
}

NodeId Graph::matmul(NodeId a, NodeId b) {
    const auto& sa = nodes_[a].shape;
    const auto& sb = nodes_[b].shape;
    if (sa.size() != 2 || sb.size() != 2) {
        throw std::invalid_argument("Graph::matmul: both operands must be rank-2");
    }
    if (sa[1] != sb[0]) {
        std::ostringstream os;
        os << "Graph::matmul: inner dim mismatch (" << sa[1] << " vs " << sb[0] << ")";
        throw std::invalid_argument(os.str());
    }
    Node n;
    n.op = OpKind::MatMul;
    n.inputs = {a, b};
    n.shape = {sa[0], sb[1]};
    return push_raw(std::move(n));
}

#define UNARY_SAME_SHAPE_IMPL(Name, Kind) \
    NodeId Graph::Name(NodeId a) { \
        Node n; n.op = OpKind::Kind; n.inputs = {a}; n.shape = nodes_[a].shape; \
        return push_raw(std::move(n)); \
    }

UNARY_SAME_SHAPE_IMPL(relu, Relu)
UNARY_SAME_SHAPE_IMPL(step, Step)
UNARY_SAME_SHAPE_IMPL(tanh, Tanh)
UNARY_SAME_SHAPE_IMPL(sigmoid, Sigmoid)
UNARY_SAME_SHAPE_IMPL(exp, Exp)
UNARY_SAME_SHAPE_IMPL(log, Log)
UNARY_SAME_SHAPE_IMPL(sqrt, Sqrt)
UNARY_SAME_SHAPE_IMPL(abs, Abs)

#undef UNARY_SAME_SHAPE_IMPL

NodeId Graph::sum(NodeId a) {
    Node n; n.op = OpKind::Sum; n.inputs = {a}; n.shape = {};
    return push_raw(std::move(n));
}

NodeId Graph::sum_axis(NodeId a, size_t axis) {
    const auto& sa = nodes_[a].shape;
    if (axis >= sa.size()) throw std::invalid_argument("Graph::sum_axis: axis out of range");
    Node n;
    n.op = OpKind::SumAxis;
    n.inputs = {a};
    n.axis = axis;
    for (size_t i = 0; i < sa.size(); ++i) if (i != axis) n.shape.push_back(sa[i]);
    return push_raw(std::move(n));
}

NodeId Graph::transpose(NodeId a) {
    const auto& sa = nodes_[a].shape;
    if (sa.size() != 2) throw std::invalid_argument("Graph::transpose: only rank-2 supported");
    Node n;
    n.op = OpKind::Transpose;
    n.inputs = {a};
    n.shape = {sa[1], sa[0]};
    return push_raw(std::move(n));
}

NodeId Graph::reshape(NodeId a, std::vector<size_t> shape) {
    const auto& sa = nodes_[a].shape;
    size_t old_n = 1; for (auto d : sa) old_n *= d;
    size_t new_n = 1; for (auto d : shape) new_n *= d;
    if (old_n != new_n) throw std::invalid_argument("Graph::reshape: element count mismatch");
    Node n;
    n.op = OpKind::Reshape;
    n.inputs = {a};
    n.target_shape = shape;
    n.shape = shape;
    return push_raw(std::move(n));
}


namespace {
struct ConvShapeOut { size_t C, OH, OW; };
ConvShapeOut conv_shape_out(const std::vector<size_t>& x_shape, size_t kh, size_t kw, size_t stride, size_t pad) {
    if (x_shape.size() != 3) throw std::invalid_argument("im2col/maxpool: expected rank-3 [C,H,W] input");
    size_t C = x_shape[0], H = x_shape[1], W = x_shape[2];
    if (H + 2 * pad < kh || W + 2 * pad < kw)
        throw std::invalid_argument("im2col/maxpool: kernel larger than padded input");
    size_t OH = (H + 2 * pad - kh) / stride + 1;
    size_t OW = (W + 2 * pad - kw) / stride + 1;
    return {C, OH, OW};
}
}

NodeId Graph::im2col(NodeId x, size_t kh, size_t kw, size_t stride, size_t pad) {
    ConvShapeOut o = conv_shape_out(nodes_[x].shape, kh, kw, stride, pad);
    Node n;
    n.op = OpKind::Im2Col;
    n.inputs = {x};
    n.shape = {o.C * kh * kw, o.OH * o.OW};
    n.kernel_h = kh; n.kernel_w = kw; n.stride = stride; n.pad = pad;
    return push_raw(std::move(n));
}

NodeId Graph::col2im(NodeId cols, size_t C, size_t H, size_t W, size_t kh, size_t kw, size_t stride, size_t pad) {
    ConvShapeOut o = conv_shape_out({C, H, W}, kh, kw, stride, pad);
    const auto& sc = nodes_[cols].shape;
    if (sc != std::vector<size_t>{C * kh * kw, o.OH * o.OW}) {
        throw std::invalid_argument("Graph::col2im: cols shape does not match expected im2col output shape");
    }
    Node n;
    n.op = OpKind::Col2Im;
    n.inputs = {cols};
    n.shape = {C, H, W};
    n.kernel_h = kh; n.kernel_w = kw; n.stride = stride; n.pad = pad;
    return push_raw(std::move(n));
}

NodeId Graph::maxpool(NodeId x, size_t ph, size_t pw, size_t stride) {
    ConvShapeOut o = conv_shape_out(nodes_[x].shape, ph, pw, stride, 0);
    Node n;
    n.op = OpKind::MaxPool;
    n.inputs = {x};
    n.shape = {o.C, o.OH, o.OW};
    n.kernel_h = ph; n.kernel_w = pw; n.stride = stride; n.pad = 0;
    return push_raw(std::move(n));
}

NodeId Graph::maxpool_backward(NodeId x, NodeId y, NodeId gy, size_t ph, size_t pw, size_t stride) {
    if (nodes_[y].shape != nodes_[gy].shape) {
        throw std::invalid_argument("Graph::maxpool_backward: y and gy shape mismatch");
    }
    Node n;
    n.op = OpKind::MaxPoolBackward;
    n.inputs = {x, y, gy};
    n.shape = nodes_[x].shape;
    n.kernel_h = ph; n.kernel_w = pw; n.stride = stride;
    return push_raw(std::move(n));
}


NodeId Graph::conv2d(NodeId x, NodeId kernel, size_t stride, size_t pad) {
    const auto& xs = nodes_[x].shape;
    const auto& ks = nodes_[kernel].shape;
    if (xs.size() != 3) throw std::invalid_argument("Graph::conv2d: x must be rank-3 [C,H,W]");
    if (ks.size() != 4) throw std::invalid_argument("Graph::conv2d: kernel must be rank-4 [OutC,C,KH,KW]");
    size_t C = xs[0];
    size_t OutC = ks[0], KC = ks[1], KH = ks[2], KW = ks[3];
    if (KC != C) throw std::invalid_argument("Graph::conv2d: kernel channel count must match x's channels");

    ConvShapeOut o = conv_shape_out(xs, KH, KW, stride, pad);
    NodeId cols = im2col(x, KH, KW, stride, pad);
    NodeId kernel2d = reshape(kernel, {OutC, C * KH * KW});
    NodeId prod = matmul(kernel2d, cols);
    return reshape(prod, {OutC, o.OH, o.OW});
}

NodeId Graph::softmax(NodeId x) {


    NodeId e = exp(x);
    NodeId s = sum(e);
    NodeId s_b = broadcast_to(s, nodes_[x].shape);
    return div(e, s_b);
}

NodeId Graph::cross_entropy(NodeId pred, NodeId target) {
    NodeId lp = log(pred);
    NodeId prod = mul(target, lp);
    NodeId s = sum(prod);
    return neg(s);
}

NodeId Graph::batch_norm(NodeId x, NodeId gamma, NodeId beta, double eps) {


    std::vector<size_t> xs = nodes_[x].shape;
    size_t n_elems = 1; for (auto d : xs) n_elems *= d;
    NodeId inv_n = constant(1.0 / static_cast<double>(n_elems), {});

    NodeId sum_x = sum(x);
    NodeId mean = mul(sum_x, inv_n);
    NodeId mean_b = broadcast_to(mean, xs);
    NodeId centered = sub(x, mean_b);
    NodeId sq = mul(centered, centered);
    NodeId sum_sq = sum(sq);
    NodeId var = mul(sum_sq, inv_n);
    NodeId var_eps = add(var, constant(eps, {}));
    NodeId std_dev = sqrt(var_eps);
    NodeId std_b = broadcast_to(std_dev, xs);
    NodeId normalized = div(centered, std_b);
    NodeId scaled = mul(gamma, normalized);
    return add(scaled, beta);
}

}
